subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/CollectDirectClassSuperTypesTransform.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.initialization.transform;
+
+import com.google.common.collect.Ordering;
+import org.gradle.api.artifacts.transform.InputArtifact;
+import org.gradle.api.artifacts.transform.TransformAction;
+import org.gradle.api.artifacts.transform.TransformOutputs;
+import org.gradle.api.artifacts.transform.TransformParameters;
+import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.internal.classpath.ClasspathWalker;
+import org.gradle.internal.file.FileException;
+import org.gradle.internal.file.Stat;
+import org.gradle.work.DisableCachingByDefault;
+import org.objectweb.asm.ClassReader;
+
+import javax.inject.Inject;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.ImmutableSortedSet.toImmutableSortedSet;
+
+/**
+ * TODO: This class has similar implementation in build-logic/packaging/src/main/kotlin/gradlebuild/instrumentation/transforms/CollectDirectClassSuperTypesTransform.kt.
+ *  We could reuse the same class at some point.
+ */
+@DisableCachingByDefault(because = "Not worth caching.")
+public abstract class CollectDirectClassSuperTypesTransform implements TransformAction<TransformParameters.None> {
+
+    private static final Predicate<String> ACCEPTED_TYPES = type -> type != null && !type.startsWith("java/lang");
+    public static final String SUPER_TYPES_MARKER_FILE_NAME = ".gradle-super-types.marker";
+    private static final String FILE_SUFFIX = ".super-types";
+
+    @Inject
+    public abstract ObjectFactory getObjects();
+
+    @PathSensitive(PathSensitivity.NAME_ONLY)
+    @InputArtifact
+    public abstract Provider<FileSystemLocation> getInput();
+
+    public void transform(TransformOutputs outputs) {
+        try {
+            InjectedInstrumentationServices services = getObjects().newInstance(InjectedInstrumentationServices.class);
```

Comment:
❓ This is because we can't directly inject those services? If so, we should add a comment here to explain why we use this hack.

Reply:
Yes, that is the case. Maybe we should just allow injecting services with some internal annotation, although probably users will abuse it then :).

Added a comment